pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -0,0 +1,235 @@
+		},
+		labels,
+	)
+	sparkAppRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix, "spark_app_running_count"),
+		"Spark App Running Count via the Operator", labels)
+	sparkAppExecutorRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix,
+		"spark_app_executor_running_count"), "Spark App Running Executor Count via the Operator", labels)
+
+	util.RegisterMetric(sparkAppSubmitCount)
+	util.RegisterMetric(sparkAppSuccessCount)
+	util.RegisterMetric(sparkAppFailureCount)
+	util.RegisterMetric(sparkAppSuccessExecutionTime)
+	util.RegisterMetric(sparkAppFailureExecutionTime)
+	util.RegisterMetric(sparkAppExecutorFailureCount)
+	util.RegisterMetric(sparkAppExecutorSuccessCount)
+
+	return &sparkAppMetrics{
+		labels,
+		prefix,
+		sparkAppSubmitCount,
+		sparkAppSuccessCount,
+		sparkAppFailureCount,
+		sparkAppRunningCount,
+		sparkAppSuccessExecutionTime,
+		sparkAppFailureExecutionTime,
+		sparkAppExecutorRunningCount,
+		sparkAppExecutorFailureCount,
+		sparkAppExecutorSuccessCount,
+	}
+}
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
+	// Transform spec labels since our labels names might be not same as specLabels if we removed invalid characters.
+	validSpecLabels := make(map[string]string)
+	for labelKey, v := range specLabels {
+		newKey := util.CreateValidMetricNameLabel("", labelKey)
+		validSpecLabels[newKey] = v
+	}
+
+	metricLabels := make(map[string]string)
+	for _, label := range labels {
+		if value, ok := validSpecLabels[label]; ok {
+			metricLabels[label] = value
+		} else {
+			metricLabels[label] = "Unknown"
+		}
+	}
+	return metricLabels
+}
+
+func (sm *sparkAppMetrics) exportMetrics(oldApp, newApp *v1alpha1.SparkApplication) {
+	metricLabels := fetchMetricLabels(newApp.Labels, sm.labels)
+	glog.V(2).Infof("Exporting Metrics for %s. OldStatus: %v NewStatus: %v", newApp.GetName(), oldApp.Status, newApp.Status)
+
+	oldState := oldApp.Status.AppState.State
+	newState := newApp.Status.AppState.State
+	switch newState {
+	case v1alpha1.SubmittedState:
+		if isValidDriverStateTransition(oldState, newState) {
+			if m, err := sm.sparkAppSubmitCount.GetMetricWith(metricLabels); err != nil {
+				glog.Errorf("Error while exporting metrics: %v", err)
+			} else {
+				m.Inc()
+			}
+		}
+	case v1alpha1.RunningState:
+		if isValidDriverStateTransition(oldState, newState) {
+			sm.sparkAppRunningCount.Inc(metricLabels)
+		}
+	case v1alpha1.CompletedState:
+		if isValidDriverStateTransition(oldState, newState) {
+			if !newApp.Status.SubmissionTime.Time.IsZero() && !newApp.Status.CompletionTime.Time.IsZero() {
+				d := newApp.Status.CompletionTime.Time.Sub(newApp.Status.SubmissionTime.Time)
+
+				if m, err := sm.sparkAppSuccessExecutionTime.GetMetricWith(metricLabels); err != nil {
+					glog.Errorf("Error while exporting metrics: %v", err)
+				} else {
+					m.Observe(float64(d / time.Microsecond))
```

Comment:
Can we do `float64(d) / time. Microsecond` instead here?

Reply:
same as above